src/erasure-code/isa/ErasureCodeIsa.cc

```diff
@@ -114,29 +114,111 @@ int ErasureCodeIsa::decode_chunks(const set<int> &want_to_read,
   return isa_decode(erasures, data, coding, blocksize);
 }
 
+void ErasureCodeIsaDefault::encode_delta(const bufferptr &old_data,
+                                  const bufferptr &new_data,
+                                  bufferptr *delta)
+{
+  
+  //TODO handle old_data or new_data being a special zero buffer.
+  
+  constexpr int data_vectors = 2;
+  char * data[data_vectors];
+  data[0] = const_cast<char*>(old_data.c_str());
+  data[1] = const_cast<char*>(new_data.c_str());
+  char * coding = delta->c_str();
+  //TODO we need to be able to handle the delta buffer being the same
+  // buffer as the old_data buffer. Call isa_xor_with_new_coding_buf
+  isa_xor(data, &coding, delta->length(), data_vectors);
+  
+  //TODO return a special zero buffer if delta is all zeroes
+}
+
+void ErasureCodeIsaDefault::apply_delta(const map<int, bufferptr> &in,
+                                        map <int, bufferptr> &out)
+{
+  auto first = in.begin();
+  const unsigned blocksize = first->second.length();
+
+  //for each delta chunk in the in map
+  for (auto i = in.begin(); i != in.end(); i++) {
+    if (i->first < k) {
+      //for each coding chunk in the out map
+      for (auto j = out.begin(); j != out.end(); j++) {
+        if (j->first >= k) {
+          if (j->second.length() < blocksize) {
+            //assert?
+            continue;
+          }
+          if (m==1) {
+            int data_vectors = 1;
+            char * input_data = const_cast<char*>(i->second.c_str());
+            char * output_data = const_cast<char*>(j->second.c_str());
+            isa_xor_with_new_coding_buf(&input_data, &output_data, blocksize, data_vectors);
+          }
+          else {
+            unsigned char* input_data = reinterpret_cast<unsigned char*>(const_cast<char*>(i->second.c_str()));
+            unsigned char* output_data = reinterpret_cast<unsigned char*>(const_cast<char*>(j->second.c_str()));
+            ec_encode_data_update(blocksize, k, 1, i->first, encode_tbls + (32 * k * (j->first - k)), input_data, &output_data);
+          }
+          //TODO return a special zero buffer if coding buffer is all zeroes
+        }
+      }
+    }
+  }
+}
+    
 // -----------------------------------------------------------------------------
 
+/* 
+ * isa_xor_with_new_coding_buf() must be used if a new empty coding buffer
+ * is needed, because xor_gen() requires an empty coding buffer.
+ * Creates a new aligned buffer to use for the coding calculation, then overwrites
+ * the caller's coding buffer with the new coding value.
+*/
 void
-ErasureCodeIsa::isa_xor(char **data, char **coding, int blocksize)
+ErasureCodeIsa::isa_xor_with_new_coding_buf(char **data, char **coding, int blocksize, int data_vectors)
```

Comment:
Why char **coding ? This is only passing in a single pointer and the function isn't modifying the pointer.